drivers/i2c/i2c_mchp_mss.c

```diff
@@ -0,0 +1,417 @@
+#define CTRL_CR1  BIT(1)
+#define CTRL_AA   BIT(2)
+#define CTRL_SI   BIT(3)
+#define CTRL_STO  BIT(4)
+#define CTRL_STA  BIT(5)
+#define CTRL_ENS1 BIT(6)
+#define CTRL_CR2  BIT(7)
+
+#define STATUS_M_START_SENT                  (0x08)
+#define STATUS_M_REPEATED_START_SENT         (0x10)
+#define STATUS_M_SLAW_ACK                    (0x18)
+#define STATUS_M_SLAW_NACK                   (0x20)
+#define STATUS_M_TX_DATA_ACK                 (0x28)
+#define STATUS_M_TX_DATA_NACK                (0x30)
+#define STATUS_M_ARB_LOST                    (0x38)
+#define STATUS_M_SLAR_ACK                    (0x40)
+#define STATUS_M_SLAR_NACK                   (0x48)
+#define STATUS_M_RX_DATA_ACKED               (0x50)
+#define STATUS_M_RX_DATA_NACKED              (0x58)
+#define STATUS_S_SLAW_ACKED                  (0x60)
+#define STATUS_S_ARB_LOST_SLAW_ACKED         (0x68)
+#define STATUS_S_GENERAL_CALL_ACKED          (0x70)
+#define STATUS_S_ARB_LOST_GENERAL_CALL_ACKED (0x78)
+#define STATUS_S_RX_DATA_ACKED               (0x80)
+#define STATUS_S_RX_DATA_NACKED              (0x88)
+#define STATUS_S_GENERAL_CALL_RX_DATA_ACKED  (0x90)
+#define STATUS_S_GENERAL_CALL_RX_DATA_NACKED (0x98)
+#define STATUS_S_RX_STOP                     (0xA0)
+#define STATUS_S_SLAR_ACKED                  (0xA8)
+#define STATUS_S_ARB_LOST_SLAR_ACKED         (0xB0)
+#define STATUS_S_TX_DATA_ACK                 (0xB8)
+#define STATUS_S_TX_DATA_NACK                (0xC0)
+#define STATUS_LAST_DATA_ACK                 (0xC8)
+
+#define PCLK_DIV_960 (CTRL_CR2)
+#define PCLK_DIV_256 (0)
+#define PCLK_DIV_224 (CTRL_CR0)
+#define PCLK_DIV_192 (CTRL_CR1)
+#define PCLK_DIV_160 (CTRL_CR0 | CTRL_CR1)
+#define PCLK_DIV_120 (CTRL_CR0 | CTRL_CR2)
+#define PCLK_DIV_60  (CTRL_CR1 | CTRL_CR2)
+#define BCLK_DIV_8   (CTRL_CR0 | CTRL_CR1 | CTRL_CR2)
+#define CLK_MASK     (CTRL_CR0 | CTRL_CR1 | CTRL_CR2)
+
+/* -- Transactions types -- */
+#define NO_TRANSACTION                     (0x00)
+#define CONTROLLER_WRITE_TRANSACTION       (0x01)
+#define CONTROLLER_READ_TRANSACTION        (0x02)
+#define CONTROLLER_RANDOM_READ_TRANSACTION (0x03)
+#define WRITE_TARGET_TRANSACTION           (0x04)
+#define READ_TARGET_TRANSACTION            (0x05)
+
+#define MSS_I2C_RELEASE_BUS (0x00)
+#define MSS_I2C_HOLD_BUS    (0x01)
+#define TARGET_ADDR_SHIFT	(0x01)
+
+typedef void (*mss_i2c_cfg_func_t)(void);
+
+enum i2c_status {
+	MSS_I2C_SUCCESS = 0u,
+	MSS_I2C_IN_PROGRESS,
+	MSS_I2C_FAILED,
+	MSS_I2C_TIMED_OUT
+};
+
+struct mss_i2c_config {
+	uint32_t clock_freq;
+	uintptr_t i2c_base_addr;
+	uint32_t i2c_irq_base;
+	void (*irq_config_func)(void);
+};
+
+struct mss_i2c_data {
+	uint8_t ser_address;
+	uint8_t target_addr;
+	uint8_t options;
+	uint8_t transaction;
+	const uint8_t *controller_tx_buffer;
+	uint16_t controller_tx_size;
+	uint16_t controller_tx_idx;
+	uint8_t dir;
+	uint8_t *controller_rx_buffer;
+	uint16_t controller_rx_size;
+	uint16_t controller_rx_idx;
+
+	volatile enum i2c_status controller_status;
```

Comment:
`atomic_t` instead?

Reply:
@carlocaione thanks, switched to `atomic_t`